numpy/core/src/umath/ufunc_object.c

```diff
@@ -1748,10 +1763,148 @@ make_arr_prep_args(npy_intp nin, PyObject *args, PyObject *kwds)
     }
 }
 
+/*
+ * Check whether any of the outputs of a gufunc has core dimensions.
+ */
+static int
+_has_output_coredims(PyUFuncObject *ufunc) {
+    int i;
+    for (i = ufunc->nin; i < ufunc->nin + ufunc->nout; ++i) {
+        if (ufunc->core_num_dims[i] > 0) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/*
+ * Interpret a possible axes keyword argument, using it to remap the default
+ * axes (at the end of the operands) to new ones.
+ *
+ * Returns 0 on success, and -1 on failure
+ */
+static int
+_parse_axes_arg(PyUFuncObject *ufunc, PyObject *axes, PyArrayObject **op,
+                int broadcast_ndim, int **remap_axis) {
+    int nin = ufunc->nin;
+    int nout = ufunc->nout;
+    int nop = nin + nout;
+    int i, list_size;
+    PyObject *item;
+
+    if (!PyList_Check(axes)) {
+        PyErr_SetString(PyExc_TypeError, "axes should be a list.");
+        return -1;
+    }
+    list_size = PyList_Size(axes);
+    if (list_size != nop) {
+        if (list_size != nin || _has_output_coredims(ufunc)) {
+            PyErr_Format(PyExc_ValueError,
+                         "axes should be a list with an entry for all "
+                         "%d inputs and outputs; entries for outputs can only "
+                         "be omitted if none of them has core axes.",
+                         nop);
+            return -1;
+        }
+        for (i = nin; i < nop; i++) {
+            remap_axis[i] = NULL;
+        }
+    }
+    for (i = 0; i < list_size; ++i) {
+        int j, iter_axis;
+        int op_ndim, core_ndim;
+        int requested_axis = 0;
+        int have_seen_axis[NPY_MAXDIMS] = {0};
+
+        core_ndim = ufunc->core_num_dims[i];
+        if (op[i] != NULL) {
+            op_ndim = PyArray_NDIM(op[i]);
+        }
+        else {
+            op_ndim = broadcast_ndim + core_ndim;
+        }
+        item = PyList_GetItem(axes, i);
+        /* if not a tuple already, make it one; content is checked later */
+        if (!PyTuple_Check(item)) {
+            item = PyTuple_Pack(1, item);
+            if (item == NULL) {
+                return -1;
+            }
+        }
+        else {
+            Py_INCREF(item);
+        }
+        if (PyTuple_Size(item) != core_ndim) {
+            if (core_ndim == 1) {
+                PyErr_Format(PyExc_TypeError,
+                             "axes item %d should be a tuple with a "
+                             "single element, or an integer", i);
+            }
+            else {
+                PyErr_Format(PyExc_TypeError,
```

Comment:
I'd argue this should be `ValueError` to match the error given by `a, b = (1,)`, which is also due to the wrong number of items.

Reply:
Done

Reply:
Actually, an annoyance is that now everything becomes a ValueError, since if one doesn't put in a tuple, things get automatically stuffed into a tuple.

Reply:
Doesn't a `TypeError` then get raised by the conversion to `int`, which is fine?

Reply:
Though I guess it is really wrong types of items in a list, so perhaps `ValueError` is OK...

Reply:
Oh, I see - perhaps then it should be
```
if (PyTuple_Check(item)) {
    Py_INCREF(item);
}
else if (core_ndim == 1) {
    item = PyTuple_Pack(1, item);  // and null check
}
else {
    raise TypeError
}
```